### PR TITLE
feat: implement eth_getStorageValues RPC method

### LINKED ISSUE #10282: implement new RPC method eth_getStorageValues
spec https://github.com/ethereum/execution-apis/issues/752
geth impl https://github.com/ethereum/go-ethereum/pull/32591

Acceptance criteria: passing these `generic` [hive](https://hive.ethpandaops.io/#/test/generic/1776665467-b060efaa21348b473929b152e261f985) tests (besu failing with no such method error)

eth_getStorageValues/get-storage-values-empty-request 
eth_getStorageValues/get-storage-values-multiple-addresses
eth_getStorageValues/get-storage-values-unknown-account

### PR DESCRIPTION
**Pull request by vivek-0509**
Added the `eth_getStorageValues` JSON-RPC method, a batched version of `eth_getStorageAt` that reads multiple storage slots across multiple accounts in a single call.

### What changed

- New `EthGetStorageValues` method extending `AbstractBlockParameterOrBlockHashMethod`
- New `StorageSlotsRequest` parameter class (`HashMap<Address, List<String>>`) 
- Registered in `RpcMethod` enum and `EthJsonRpcMethods`
- Unit tests covering single address, multiple addresses, unknown accounts, empty request, and the slot cap

## Fixed Issue(s)

Fixes 4 failing Hive rpc-compat tests for `eth_getStorageValues`:
https://hive.ethpandaops.io/#/test/generic/1776233102-b0ae09a940116cc8abb883b9505fa1e9
fixes #10282 

### Thanks for sending a pull request! Have you done the following?

- [x] Checked out our [contribution guidelines](https://github.com/besu-eth/besu/blob/main/CONTRIBUTING.md)?
- [ ] Considered documentation and added the `doc-change-required` label to this PR [if updates are required](https://wiki.hyperledger.org/display/BESU/Documentation).
- [x] Considered the changelog and included an [update if required](https://wiki.hyperledger.org/display/BESU/Changelog).
- [ ] For database changes (e.g. KeyValueSegmentIdentifier) considered compatibility and performed forwards and backwards compatibility tests

### Locally, you can run these tests to catch failures early:

- [x] spotless: `./gradlew spotlessApply`
- [x] unit tests: `./gradlew build`
- [ ] acceptance tests: `./gradlew acceptanceTest`
- [ ] integration tests: `./gradlew integrationTest`
- [ ] reference tests: `./gradlew ethereum:referenceTests:referenceTests`
- [x] hive tests: [Engine or other RPCs modified?](https://lf-hyperledger.atlassian.net/wiki/spaces/BESU/pages/22156302/Using+Hive+Test+Suite)

### CODE REVIEW

**Comment by vivek-0509** [comment]
@fab-10 PR is ready for review please take a look

**Comment by vivek-0509** [comment]
@fab-10 merged with main

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetStorageValuesTest.java [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    final Map<String, List<String>> result = (Map<String, List<String>>) response.getResult();
+    assertThat(result)
+        .containsEntry(ADDRESS_ONE.toHexString(), List.of(VALUE_FORTY_TWO))
+        .containsEntry(ADDRESS_TWO.toHexString(), List.of(VALUE_ZERO));
+  }
+
+  @Test
+  public void shouldReturnZeroForUnknownAccount() {
+    when(blockchainQueries.storageAt(ADDRESS_TWO, UInt256.fromHexString(SLOT_HIGH), Hash.ZERO))
+        .thenReturn(Optional.of(UInt256.ZERO));
```
Comment (macfarla) [inline ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetStorageValuesTest.java]:
```suggestion
        .thenReturn(Optional.empty());
```

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetStorageValues.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+          new JsonRpcError(RpcErrorType.INVALID_PARAMS.getCode(), "empty request", null));
+    }
+
+    final Map<String, List<String>> result = new HashMap<>();
+    for (final Map.Entry<Address, List<String>> entry : storageSlotsRequest.entrySet()) {
+      final Address address = entry.getKey();
+      final List<String> keys = entry.getValue();
+      final List<String> values = new ArrayList<>(keys.size());
+      for (final String keyHex : keys) {
+        final UInt256 key = UInt256.fromHexString(keyHex);
```
Comment (macfarla) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetStorageValues.java]: Malformed slot hex (e.g. "0xzz") will throw an uncaught IllegalArgumentException, resulting in an unhandled 500-style error instead of a clean INVALID_PARAMS response. Should wrap in try/catch.

**Review by macfarla — COMMENTED** [review]
couple of comments.

confirming I have run the rpc-compat hive tests locally and the new eth_getStorageAt tests pass

eth_getStorageValues/get-storage-values-empty-request (besu_besu-local) | ✓ | 3ms |  
-- | -- | -- | --
eth_getStorageValues/get-storage-values-multiple-addresses (besu_besu-local) | ✓ | 2ms |  
eth_getStorageValues/get-storage-values-unknown-account (besu_besu-local) | ✓ | 2ms |  
eth_getStorageValues/get-storage-values (besu_besu-local)

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetStorageValuesTest.java (hunk 1 above)
Comment (vivek-0509) [reply ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetStorageValuesTest.java]: done

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetStorageValues.java (hunk 2 above)
Comment (vivek-0509) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetStorageValues.java]: done

**Comment by vivek-0509** [comment]
@macfarla  all changes done

File: CHANGELOG.md [hunk 3]
```diff
@@ -61,6 +61,7 @@ are provided with different values, using input as per the execution-apis spec i
 - Upgrade besu-native libraries version to 1.5.0. This fixes the issue of besu-native/secp256r1 exporting OpenSSL symbols in JVM space. [besu-native #308](https://github.com/besu-eth/besu-native/pull/308)
 
 ### Additions and Improvements
+- Add `eth_getStorageValues` JSON-RPC method for batched reads of multiple storage slots across multiple accounts in a single call [#10259](https://github.com/besu-eth/besu/pull/10259)
```
Comment (macfarla) [inline CHANGELOG.md]: can you move the changelog entry to the Unreleased section

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetStorageValues.java (hunk 2 above)
Comment (macfarla) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetStorageValues.java]: can you add test coverage for this case

**Review by macfarla — COMMENTED** [review]
looking good, couple of comments

File: CHANGELOG.md (hunk 3 above)
Comment (vivek-0509) [reply CHANGELOG.md]: done

**Comment by vivek-0509** [comment]
@macfarla all done

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetStorageValues.java (hunk 2 above)
Comment (vivek-0509) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetStorageValues.java]: done

**Review by macfarla — COMMENTED** [review]
@vivek-0509 can you update with latest main and resolve the conflict in changelog

assuming you've run the tests locally I think it's pretty close. once you've updated I'll re-review

**Comment by vivek-0509** [comment]
@macfarla done, build passes locally

**Comment by fab-10** [comment]
Please add also some integration test scenarios to `EthJsonRpcHttpBySpecTest`, the json files are defined in `ethereum/api/src/test/resources/org/hyperledger/besu/ethereum/api/jsonrpc/eth` see there for examples

**Comment by vivek-0509** [comment]
@fab-10 @macfarla all changes done please take a look

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetStorageValuesTest.java [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    final JsonRpcSuccessResponse response = (JsonRpcSuccessResponse) method.response(request);
+
+    @SuppressWarnings("unchecked")
+    final Map<String, List<String>> result = (Map<String, List<String>>) response.getResult();
+    assertThat(result).containsEntry(ADDRESS_TWO.toHexString(), List.of(VALUE_ZERO));
+  }
+
+  @Test
+  public void shouldReturnTooManySlotsErrorWhenOver1024() {
+    final String[] keys = new String[1025];
```
Comment (macfarla) [inline ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetStorageValuesTest.java]: better to use MAX+1 rather than hardcoding 1025

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetStorageValues.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    final StorageSlotsRequest storageSlotsRequest;
+    try {
+      storageSlotsRequest = request.getRequiredParameter(0, StorageSlotsRequest.class);
+    } catch (JsonRpcParameterException e) {
+      throw new InvalidJsonRpcParameters(
+          "Invalid storage slots request parameter (index 0)", RpcErrorType.INVALID_PARAMS, e);
+    }
+
+    int totalSlots = 0;
+    for (final List<String> keys : storageSlotsRequest.values()) {
```
Comment (macfarla) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetStorageValues.java]: Bug: NPE if a slot list is null in the JSON

  If the request JSON contains {"0xaddr": null}, Jackson will insert a null value into the StorageSlotsRequest map. Then:
  for (final List<String> keys : storageSlotsRequest.values()) {
      totalSlots += keys.size();  // NullPointerException
  }
  Should either guard against null values in the iteration, or annotate the map values with @JsonSetter(nulls = Nulls.SKIP) / handle in deserialization.

  Similarly, a null element inside a slot list (["0x0", null, "0x1"]) would cause a NPE in UInt256.fromHexString(null) which is not caught by the existing IllegalArgumentException catch.

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetStorageValuesTest.java [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import org.apache.tuweni.units.bigints.UInt256;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+public class EthGetStorageValuesTest {
```
Comment (macfarla) [inline ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetStorageValuesTest.java:48]: The block hash path is only exercised by JSON fixtures, not by the unit test class. The unit tests only use "latest". Worth adding a unit test for resultByBlockHash with a specific block hash to complement the mock setup.

**Review by macfarla — COMMENTED** [review]
overall looking good. some comments but it's getting close

**Comment by vivek-0509** [comment]
@macfarla  all done build passes locally

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetStorageValuesTest.java (hunk 4 above)
Comment (vivek-0509) [reply ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetStorageValuesTest.java]: done

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetStorageValues.java (hunk 5 above)
Comment (vivek-0509) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetStorageValues.java]: done

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetStorageValuesTest.java (hunk 6 above)
Comment (vivek-0509) [reply ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetStorageValuesTest.java:48]: done

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/parameters/StorageSlotsRequest.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+import org.hyperledger.besu.datatypes.Address;
+
+import java.util.HashMap;
+import java.util.List;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+
+/** Map of storage slot requests, indexed by address. */
+@JsonIgnoreProperties(ignoreUnknown = true)
```
Comment (macfarla) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/parameters/StorageSlotsRequest.java]: @JsonIgnoreProperties(ignoreUnknown = true) is ineffective on a HashMap subclass

For a Map type, Jackson treats all JSON object properties as map entries — there are no "unknown" properties to ignore. The annotation has no effect and should be removed.

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetStorageValues.java [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.tuweni.units.bigints.UInt256;
+
+public class EthGetStorageValues extends AbstractBlockParameterOrBlockHashMethod {
+
+  static final int MAX_STORAGE_SLOTS = 1024;
```
Comment (macfarla) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetStorageValues.java:43]: Where does 1024 come from? If this matches a specific spec or Geth behaviour, add a comment. If it's arbitrary, that's also worth documenting. A magic constant without context is a maintenance hazard.

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/parameters/StorageSlotsRequest.java (hunk 7 above)
Comment (vivek-0509) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/parameters/StorageSlotsRequest.java]: done

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetStorageValues.java (hunk 8 above)
Comment (vivek-0509) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetStorageValues.java:43]: Done. The 1024 comes from geth's reference implementation in [go-ethereum#32591](https://github.com/ethereum/go-ethereum/pull/32591), i have kept the same value for spec parity, since the cap is enforced the same way (total slots across all addresses, error if exceeded). Added a comment citing geth as the source.

**Comment by vivek-0509** [comment]
@macfarla all changes done

**Comment by vivek-0509** [comment]
@macfarla while addressing the two open comments, I noticed a performance issue I wanted to flag and i have fixed in the same push. The current implementation calls blockchainQueries.storageAt(addr, key, blockHash) once per slot inside the loop. Each storageAt call internally goes through getAndMapWorldState, which opens a fresh MutableWorldState, walks the account trie, reads the slot, then closes the world state. So a request with 1024 slots opens and closes the world state 1024 times, even though all reads are at the same block hash. 

I fixed this by switching the read loop to a single blockchainQueries.getAndMapWorldState(blockHash, ws -> {}) call that iterates all (address, slot) pairs against that one world state. The hex slot keys are now pre-parsed into UInt256 during the existing validation loop, before the world state is opened, because getAndMapWorldState has a catch-all that would otherwise swallow InvalidJsonRpcParameters thrown from inside the lambda and misclassify them as worldstate failures. I also added an .orElseGet() returning a WORLD_STATE_UNAVAILABLE error for the case where state can't be opened. The current code silently returns zeros in that case, which is indistinguishable from "the slot is genuinely zero," so this is a small correctness improvement on the side. Hive tests continue to pass tested locally.

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetStorageValues.java [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+    // Validate and pre-parse all hex keys before opening the world state. Keeping validation
+    // outside the getAndMapWorldState lambda avoids its catch-all swallowing input errors.
+    final Map<Address, List<UInt256>> parsed = new HashMap<>(storageSlotsRequest.size());
+    int totalSlots = 0;
+    for (final Map.Entry<Address, List<String>> entry : storageSlotsRequest.entrySet()) {
+      final List<String> keys = entry.getValue();
+      if (keys == null) {
+        return new JsonRpcErrorResponse(
+            request.getRequest().getId(),
+            new JsonRpcError(RpcErrorType.INVALID_PARAMS.getCode(), "null slot list", null));
```
Comment (macfarla) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthGetStorageValues.java]:
```suggestion
  new JsonRpcError(RpcErrorType.INVALID_PARAMS, "null slot list")
```

**Review by macfarla — COMMENTED** [review]
EthGetStorageAt uses UInt256Parameter for slot key parsing, which handles edge cases around hex formatting consistently. This method calls UInt256.fromHexString() directly. Should match the existing pattern for consistency.

it would be good to add a test for multiple slots per address - all unit tests use exactly 1 slot per address.

**Comment by macfarla** [comment]
@vivek-0509 are you still working on this one? some final tidy up needed but it's pretty close

**Comment by vivek-0509** [comment]
@macfarla i will push the changes

**Comment by vivek-0509** [comment]
@macfarla all changes done

**Review by macfarla — APPROVED** [review]
Thanks for your contribution @vivek-0509 and your persistence in getting it across the line - approved!

**Comment by vivek-0509** [comment]
@macfarla Flaky test, unrelated to this PR. Could you re-run CI please?
